url2b521cbda7_eobermuhlner_jhuge_tgz-pJ8-ch_obermuhlner_jhuge_example_MeasureHeapJ8/src/ch/obermuhlner/jhuge/memory/MemoryMappedFileManager.java

```diff
@@ -0,0 +1,581 @@
+	 */
+	public MemoryMappedFileManager() {
+		this(100 * MEGABYTES, NO_BLOCK_SIZE);
+	}
+	
+	/**
+	 * Constructs a {@link MemoryMappedFileManager} with the specified buffer size and no block quantification.
+	 * 
+	 * @param bufferSize the buffer size 
+	 */
+	public MemoryMappedFileManager(int bufferSize) {
+		this(bufferSize, NO_BLOCK_SIZE);
+	}
+	
+	/**
+	 * Constructs a {@link MemoryMappedFileManager} with the specified buffer and block size.
+	 * 
+	 * @param bufferSize the buffer size 
+	 * @param blockSize the block size, or {@link #NO_BLOCK_SIZE} to use no block quantification
+	 */
+	public MemoryMappedFileManager(int bufferSize, int blockSize) {
+		this(bufferSize, blockSize, 32, false);
+	}
+
+	/**
+	 * Constructs a {@link MemoryMappedFileManager} with the specified configuration.
+	 * 
+	 * @param bufferSize the buffer size 
+	 * @param blockSize the block size, or {@link #NO_BLOCK_SIZE} to use no block quantification
+	 * @param allowedBlockOversize the allowed block oversize when searching for a fitting free block
+	 * @param compactAfterFree compact after freeing a block
+	 */
+	public MemoryMappedFileManager(int bufferSize, int blockSize, int allowedBlockOversize, boolean compactAfterFree) {
+		this.bufferSize = bufferSize;
+		this.blockSize = blockSize;
+		this.allowedBlockOversize = allowedBlockOversize;
+		this.compactAfterFree = compactAfterFree;
+	}
+	
+	/**
+	 * Returns the buffer size.
+	 * 
+	 * @return the buffer size
+	 */
+	public int getBufferSize() {
+		return bufferSize;
+	}
+	
+	/**
+	 * Returns the block size.
+	 * 
+	 * @return the block size, or {@link #NO_BLOCK_SIZE}
+	 */
+	public int getBlockSize() {
+		return blockSize;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 * @throws IllegalArgumentException if the specified length is greater than the buffer size
+	 */
+	@Override
+	public long allocate(int length) {
+		if (length == 0) {
+			if (emptyBlockAddress == NO_ADDRESS) {
+				emptyBlockAddress = findFreeBlock(0);
+				allocatedBlocks++;
+			}
+			if (DEBUG_TRACKING) printTrackingInfo("allocate", length);
+			return emptyBlockAddress;
+		}
+		
+		long address = findFreeBlock(quantify(length));
+		if (DEBUG) allocatedBlocksList.add(address);
+		allocatedBlocks++;
+		
+		if (DEBUG_TRACKING) printTrackingInfo("allocate", length);
+		return address;
+	}
+
+	@Override
+	public byte[] read(long address) {
+		ByteBuffer byteBuffer = readByteBufferInternal(address);
+		int length = byteBuffer.getInt();
+		if (DEBUG) checkBlockLength(address, length);
+		
+		byte[] data = new byte[length];
+		byteBuffer.get(data);
+		
+		return data;
+	}
+	
+	private ByteBuffer readByteBufferInternal(long address) {
+		if (address == emptyBlockAddress) {
+			return ByteBuffer.wrap(EMPTY_BLOCK_DATA);
+		}
+		
+		int bufferIndex = (int) (address / bufferSize);
+		int bufferPos = (int) (address % bufferSize);
+		
+		ByteBuffer buffer = buffers.get(bufferIndex);
+		buffer.position(bufferPos);
+		return buffer;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 186:195, 212:222.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.